drivers/auxdisplay/auxdisplay_itron.c

```diff
@@ -0,0 +1,378 @@
+	uint8_t brightness;
+#ifdef CONFIG_MULTITHREADING
+	struct k_sem busy_wait_sem;
+	struct gpio_callback busy_wait_callback;
+#endif
+};
+
+struct auxdisplay_itron_config {
+	const struct device *uart;
+	struct auxdisplay_capabilities capabilities;
+	struct gpio_dt_spec reset_gpio;
+	struct gpio_dt_spec busy_gpio;
+};
+
+static int send_cmd(const struct device *dev, const uint8_t *command, uint8_t length);
+static int auxdisplay_itron_is_busy(const struct device *dev);
+static int auxdisplay_itron_clear(const struct device *dev);
+
+#ifdef CONFIG_PM_DEVICE
+static int auxdisplay_itron_pm_action(const struct device *dev, enum pm_device_action action);
+#endif
+
+#ifdef CONFIG_MULTITHREADING
+void auxdisplay_itron_busy_gpio_change_callback(const struct device *port,
+						struct gpio_callback *cb,
+						gpio_port_pins_t pins)
+{
+	struct auxdisplay_itron_data *data = CONTAINER_OF(cb,
+			struct auxdisplay_itron_data, busy_wait_callback);
+	k_sem_give(&data->busy_wait_sem);
+}
+#endif
+
+static int auxdisplay_itron_init(const struct device *dev)
+{
+	const struct auxdisplay_itron_config *config = dev->config;
+	struct auxdisplay_itron_data *data = dev->data;
+	int rc;
+
+	if (!device_is_ready(config->uart)) {
+		LOG_ERR("UART device not ready");
+		return -ENODEV;
+	}
+
+	/* Configure and set busy GPIO */
+	if (config->busy_gpio.port) {
+		rc = gpio_pin_configure_dt(&config->busy_gpio, GPIO_INPUT);
+
+		if (rc < 0) {
+			LOG_ERR("Configuration of text display busy GPIO failed: %d", rc);
+			return rc;
+		}
+
+#ifdef CONFIG_MULTITHREADING
+		k_sem_init(&data->busy_wait_sem, 0, 1);
+
+		gpio_init_callback(&data->busy_wait_callback,
+				   auxdisplay_itron_busy_gpio_change_callback,
+				   BIT(config->busy_gpio.pin));
+		rc = gpio_add_callback(config->busy_gpio.port, &data->busy_wait_callback);
+
+		if (rc != 0) {
+			LOG_ERR("Configuration of busy interrupt failed: %d", rc);
+			return rc;
+		}
+#endif
+	}
+
+	/* Configure and set reset GPIO */
+	if (config->reset_gpio.port) {
+		rc = gpio_pin_configure_dt(&config->reset_gpio, GPIO_OUTPUT_INACTIVE);
+		if (rc < 0) {
+			LOG_ERR("Configuration of text display reset GPIO failed");
+			return rc;
+		}
+	}
+
+	data->character_x = 0;
+	data->character_y = 0;
+	data->brightness = 0;
+
+	/* Reset display to known configuration */
+	if (config->reset_gpio.port) {
+		uint8_t wait_loops = 0;
+
+		gpio_pin_set_dt(&config->reset_gpio, 1);
+		k_sleep(AUXDISPLAY_ITRON_RESET_TIME);
+		gpio_pin_set_dt(&config->reset_gpio, 0);
+		k_sleep(AUXDISPLAY_ITRON_RESET_WAIT_TIME);
+
+		while (auxdisplay_itron_is_busy(dev) == 1) {
+			/* Display is busy, wait */
+			k_sleep(AUXDISPLAY_ITRON_BUSY_DELAY_TIME_CHECK);
+			++wait_loops;
+
+			if (wait_loops >= AUXDISPLAY_ITRON_BUSY_WAIT_LOOPS) {
+				/* Waited long enough for display not to be busy, bailing */
+				return -EIO;
+			}
+		}
+	} else {
+#ifdef CONFIG_PM_DEVICE
+		/* Ensure display is powered on */
+		auxdisplay_itron_pm_action(dev, PM_DEVICE_ACTION_RESUME);
+#endif
```

Comment:
this call must never be invoked directly, but via e.g. runtime PM calls.

Reply:
The problem is that if the board is powered then puts the display into suspend mode, the only way to wake the screen up is to send the wake up command, which is what this does. So if you press the reset button on the board whilst the display was suspended, without this code it will not be usable because it is still in sleep mode, so the function is called directly without going through the PM system so that it just sends the wake up command. Either that or I need to duplicate the same code in 2 functions.

Reply:
why can't `pm_device_runtime_get(dev)` be called?

Reply:
The device's PM would show as active, since the board would be starting up? The issue is caused if the device is put into sleep mode and MCU is rebooted, the display keeps its sleep status but zephyr does not know that the display was suspended before it was reset.

Reply:
when using the device runtime framework, init call is expected to leave the device disabled, then, it's only activated as needed at runtime. Ref. https://docs.zephyrproject.org/latest/services/pm/device_runtime.html

Reply:
I've not used that or seen it used in any drivers, I use `pm_device_action_run`. And that functionality does not really make sense to a display since in sleep mode the display turns off, the application (not the driver) should control the power state of the display, not turning it on whilst functions are called then turning it off again after since there'd be no point in having a display. And this is how other drivers work too, in fact I can only see 10 drivers in the whole zephyr tree using `pm_device_runtime_get`